pinojs/pino
# Add formatters for core log components

Hello everyone!
This pr is the continuation of the proposal in https://github.com/pinojs/pino/issues/764.
Once I wrote the first PoC, I've changed quite a bit the approach. Following a brief explanation of what ported me to write this proposal.

As you know, one of the reasons why Pino is very fast™ is because there is a lot of caching going on, which makes impossible to have a single endpoint where a user can specify a format of the logs. Probably the best solution to this problem is https://github.com/pinojs/pino/pull/614.
The most powerful tool that Pino is offering for addressing the "format" problem, are the serializers. The only issue is that the serializers are for everything that will be passed in the pino methods, but not for the base log structure, such as the log level and all the bindings.
This proposal adds a serializer also for the level handling and the bindings, and it works as follows:
```js
const {
  bindingsSerializerSym,
  levelSerializerSym
} = require('./lib/symbols')

const pino = require('./')({
  serializers: {
    [bindingsSerializerSym] (bindings) {
      return {
        process: { pid: bindings.pid },
        hostname: bindings.hostname
      }
    },
    [levelSerializerSym] (label, number) {
      return {
        log: { level: label }
      }
    }
  }
})

pino.info('hello world')

const child = pino.child({
  foo: 'bar',
  nested: { object: true },
  serializers: {
    [bindingsSerializerSym] (bindings) {
      return { hello: 'world', ...bindings }
    }
  }
})

child.info('test')
```
The produced logs are:
```
{"log":{"level":"info"},"time":1582036600545,"process":{"pid":38765},"hostname":"Skadi.local","msg":"hello world"}
{"log":{"level":"info"},"time":1582036600546,"process":{"pid":38765},"hostname":"Skadi.local","process":{"pid":38765},"hostname":"Skadi.local","hello":"world","foo":"bar","nested":{"object":true},"msg":"test"}
```

What I like about this approach is that it's not "yet another option", but it's a feature and Pino itself is the first customer, also, the developer experience is better than configuring by hand the pieces of the JSON that needs to be assembled, like is happening for `time`.

The pull request is not yet ready to be merged, as I wanted to have an early review of the approach and see if you like it.
If so, what I think should still be done is:
- add a serializer also for timestamps
- add a serializer also for the message
- remove the multitude of options for customizing levels/timestamps/message, as the serializer would be the only way for changing these values
- provide more prepackaged serializers in `pino-std-serializers`

Following, some benchmarks:
```
=== basic.bench.js ===
# next branch
benchPino*10000: 194.908ms
benchPinoExtreme*10000: 72.674ms

# this pr
benchPino*10000: 188.696ms
benchPinoExtreme*10000: 71.806ms

=== child.bench.js ===
# next branch
benchPinoChild*10000: 217.076ms
benchPinoExtremeChild*10000: 80.304ms

# this pr
benchPinoChild*10000: 212.370ms
benchPinoExtremeChild*10000: 88.715ms
```
If this work will get accepted I want to do a follow up where users can configure the order of the keys of the JSON, it shouldn't be a problem to support it and I don't see it introducing too much overhead.
